src/_nebari/stages/infrastructure/__init__.py

```diff
@@ -554,6 +556,37 @@ def _check_input(cls, data: Any) -> Any:
                         f"Amazon Web Services instance {node_group.instance} not one of available instance types={available_instances}"
                     )
 
+        # check if kms key is valid
+        available_kms_keys = amazon_web_services.kms_key_arns(data["region"])
+        if "eks_kms_arn" in data and data["eks_kms_arn"] is not None:
+            key_id = [
+                id for id in available_kms_keys.keys() if id in data["eks_kms_arn"]
+            ]
+            if (
+                len(key_id) == 1
+                and available_kms_keys[key_id[0]]["Arn"] == data["eks_kms_arn"]
+            ):
+                key_id = key_id[0]
+                # Symmetric KMS keys with Encrypt and decrypt key-usage have the SYMMETRIC_DEFAULT key-spec
+                # EKS cluster encryption requires a Symmetric key that is set to encrypt and decrypt data
+                if available_kms_keys[key_id]["KeySpec"] != "SYMMETRIC_DEFAULT":
+                    if available_kms_keys[key_id]["KeyUsage"] == "GENERATE_VERIFY_MAC":
+                        raise ValueError(
+                            f"Amazon Web Services KMS Key with ID {key_id} does not have KeyUsage set to 'Encrypt and decrypt' data"
+                        )
+                    elif available_kms_keys[key_id]["KeyUsage"] != "ENCRYPT_DECRYPT":
+                        raise ValueError(
+                            f"Amazon Web Services KMS Key with ID {key_id} is not of type Symmetric, and KeyUsage not set to 'Encrypt and decrypt' data"
+                        )
+                    else:
+                        raise ValueError(
+                            f"Amazon Web Services KMS Key with ID {key_id} is not of type Symmetric"
+                        )
+            else:
+                raise ValueError(
+                    f"Amazon Web Services KMS Key with ARN {data['eks_kms_arn']} not one of available/enabled keys={[v['Arn'] for v in available_kms_keys.values()]}"
+                )
+
```

Comment:
You could avoid a lot of nesting here by flipping the logic on these if statements. Something like:
```python
        available_kms_keys = amazon_web_services.kms_key_arns(data["region"])
        # don't check if eks_kms_arn is not set
        if "eks_kms_arn" not in data or data["eks_kms_arn"] is None:
            return data
        key_id = [
            id for id in available_kms_keys.keys() if id in data["eks_kms_arn"]
        ]
        # Raise error if key_id is not found in available_kms_keys
        if (
            len(key_id) != 1
            or available_kms_keys[key_id[0]]["Arn"] != data["eks_kms_arn"]
        ):
            raise ValueError(
                f"Amazon Web Services KMS Key with ARN {data['eks_kms_arn']} not one of available/enabled keys={[v['Arn'] for v in available_kms_keys.values()]}"
            )

        key_id = key_id[0]
        # EKS cluster encryption requires a Symmetric key that is set to encrypt and decrypt data
        if available_kms_keys[key_id]["KeySpec"] != "SYMMETRIC_DEFAULT":
            raise ValueError(
                f"Amazon Web Services KMS Key with ID {key_id} is not of type Symmetric"
            )
        if available_kms_keys[key_id]["KeyUsage"] != "ENCRYPT_DECRYPT":
            raise ValueError(
                f"Amazon Web Services KMS Key with ID {key_id} KeyUsage not set to 'Encrypt and decrypt' data"
            )        available_kms_keys = amazon_web_services.kms_key_arns(data["region"])
        # don't check if eks_kms_arn is not set
        if "eks_kms_arn" not in data or data["eks_kms_arn"] is None:
            return data
        key_id = [
            id for id in available_kms_keys.keys() if id in data["eks_kms_arn"]
        ]
        # Raise error if key_id is not found in available_kms_keys
        if (
            len(key_id) != 1
            or available_kms_keys[key_id[0]]["Arn"] != data["eks_kms_arn"]
        ):
            raise ValueError(
                f"Amazon Web Services KMS Key with ARN {data['eks_kms_arn']} not one of available/enabled keys={[v['Arn'] for v in available_kms_keys.values()]}"
            )

        key_id = key_id[0]
        # EKS cluster encryption requires a Symmetric key that is set to encrypt and decrypt data
        if available_kms_keys[key_id]["KeySpec"] != "SYMMETRIC_DEFAULT":
            raise ValueError(
                f"Amazon Web Services KMS Key with ID {key_id} is not of type Symmetric"
            )
        if available_kms_keys[key_id]["KeyUsage"] != "ENCRYPT_DECRYPT":
            raise ValueError(
                f"Amazon Web Services KMS Key with ID {key_id} KeyUsage not set to 'Encrypt and decrypt' data"
            )
```
Has less deeply nested logic and is therefore easier to read and test.

Reply:
my hesitation to invoke `return data` in numerous places in this section is that could cause issues in the future if additional validators are appended after the KMS key validator lines under `_check_input` for aws services.

Reply:
@joneszc I can see that. It might make sense to pull every validator into its own method then, and have `_check_input` just call a list of validators. That way each could exit early if they can and nothing would get skipped. I think that would also be cleaner for readability and avoid too much responsibility for a single method.

Reply:
That would make sense.  I could open a separate PR for breaking out multiple methods for _check_input, if you'd like.  For now, I've updated the logic, based on your recommendation, for the first check.  I've also added a check to ensure the KMS key is a customer-managed versus aws-managed key.  I've left the final check in place with conditionals to precisely triage the error; for example, there is a case for a `SYMMETRIC_DEFAULT` key that could seem adequate but will fail if it is "GENERATE_VERIFY_MAC" rather than "ENCRYPT_DECRYPT".

Reply:
Can you open an issue for breaking out the _check_input methods so that doesn't get lost?

Reply:
yes!

Reply:
@viniciusdc
yes, for sure